### PR TITLE
Export benchmarking data via Prometheus exporter at endpoint /metrics

### PR DESCRIPTION
**Pull request by patsonluk**
## Description
Currently solr-bench export test results to a json file upon test completion. However, it would also be useful to have the test metrics be exposed via Prometheus /metrics endpoint such that data can be graphed by other tools such as Grafana

## Solution
Added new config parameter `prometheus-export` to the workflow (top-level) config. 

```
{
  ...
  "prometheus-export": { "port": 1234, "type-label": "facet" }
}
```

`port` is the port used for prometheus /metrics endpoint (default to 11100, take note that we picked this to avoid the default 9100 port used for prometheus node-exporter)
`type-label` is a custom label that user can define to better group metrics. default to `unknown`. There's also an override `prometheus-type-label` under `query-benchmark` in case if user wants more granular control on such label to per benchmark setup.

To enable Prometheus with defaults, simply add `"prometheus-export": {}`

`Histogram` is used to store the metrics with labels:
1. `method` : the http Method used for the call
2. `path` : the relative path used . For example `/select`, `/update` etc
3. `type` : the custom label described above. Default to `unknown`

The current implementation explicitly declares the metrics to be captured (as histogram) and only capture metrics from `QueryBenchmark` and `IndexBenchmark`. Other benchmarking tasks are not prometheus enabled ATM.
 
Take note that the Prometheus endpoint server would only be started if `prometheus-export` is defined with valid metrics to export. (ie if the benchmark only does restart, it wouldn't start the prometheus export endpoint)

## Remarks
This PR also adds some extra features, fixed some bugs:
1. `stress.sh` would now read the json config file with a new param `extra-jvm-args`, which the value will be added to the java startup call. This is useful for SSL support
2. Fixed an issue with dangling ZK client connection that prevents proper shutdown

### CODE REVIEW

**Comment by patsonluk** [comment]
@chatman Adding ability to export benchmarking metrics via grafana /metrics endpoint (different from the existing `prometheusMetrics` collector which collects prometheus metrics from the current host)

Also tagging @nginthfs and @hiteshk25 😊

**Review by nginthfs — APPROVED** [review]
Some docs could be updated but LGTM!

File: README.md [hunk 1]
```diff
@@ -87,3 +87,22 @@ Mac OS requires a few tools to run this script. Install the following:
 
 ## Acknowledgement
 This started as a project funded by Google Summer of Code (SOLR-10317), later supported by FullStory.
```
[the remark this answers is not in the document]
Comment (patsonluk) [reply README.md]: Thanks! Yes the current README lacks descriptions on the config 😓 

I have added another section for SSL access which mentions extra-jvm-args for now 😅  https://github.com/fullstorydev/solr-bench/pull/84/commits/c5abb419432c44b650f0cc0d039401f27041bace

File: src/main/java/org/apache/solr/benchmarks/query/DetailedQueryStatsListener.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  private final ConcurrentMap<String, SynchronizedDescriptiveStatistics> docHitCountStatsByType = new ConcurrentHashMap<>();
+  private final ConcurrentMap<String, AtomicLong> errorCountStatsByType = new ConcurrentHashMap<>();
+  private final AtomicBoolean loggedQueryRspError = new AtomicBoolean(false);
+  private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  @Override
+  public void onExecutionComplete(BenchmarksMain.OperationKey key, QueryResponseContents response, long duration) {
+    if (key != null) {
+      SynchronizedDescriptiveStatistics durationStats = durationStatsByType.computeIfAbsent((String) key.getAttributes().get("query"), (typeKey) -> new SynchronizedDescriptiveStatistics());
+      durationStats.addValue(duration / 1_000_000.0);
```
Comment (hiteshk25) [inline src/main/java/org/apache/solr/benchmarks/query/DetailedQueryStatsListener.java:38]: it should after successful update, below?

**Review by hiteshk25 — APPROVED** [review]
LGTM

File: src/main/java/org/apache/solr/benchmarks/beans/Cluster.java [hunk 3]
```diff
@@ -54,5 +54,8 @@ public static class ExternalSolrConfig {
 
     @JsonProperty("ssh-username")
     public String sshUserName;
+
+    @JsonProperty("protocol")
+    public String protocol = "http";
```
Comment (chatman) [inline src/main/java/org/apache/solr/benchmarks/beans/Cluster.java:59]: Makes sense for now. I'm thinking that we should convert this and others like this to Enums for better validation and readability. Maybe in a separate refactoring effort.

**Review by chatman — APPROVED** [review]
Changes look good! Thanks..

Is it possible to include a sample configuration exposing some of these changes? Not a blocker, we can do it later also.

File: src/main/java/org/apache/solr/benchmarks/query/DetailedQueryStatsListener.java (hunk 2 above)
Comment (patsonluk) [reply src/main/java/org/apache/solr/benchmarks/query/DetailedQueryStatsListener.java:38]: I think my thought was that it follows mostly the qtime logic that it handles the duration despite of the result.

I suppose there's a possibility (rather remote tho), that we DO want to benchmark on query failure? 😅 

I understand the concern is that we might overlook failures? Those should be quite obvious when doc hit count is 0 and error count is non zero. It also prints warning to the log

**Comment by patsonluk** [comment]
> Changes look good! Thanks..
> 
> Is it possible to include a sample configuration exposing some of these changes? Not a blocker, we can do it later also.

Thank you so much for the review!

As for the config, I thought of adding new one or changing existing one, but it could be a bit hard for users as it's not super trivial which config would have the settings. 

As a part of this PR, we did add description to the README.md and the sample usage: https://github.com/fullstorydev/solr-bench/pull/84/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R88 .

**Pull request merged**